src/wallet/coinselection.h

```diff
@@ -237,11 +243,37 @@ struct OutputGroup
         m_subtract_fee_outputs(params.m_subtract_fee_outputs)
     {}
 
-    void Insert(const COutput& output, size_t ancestors, size_t descendants, bool positive_only);
+    void Insert(const std::shared_ptr<COutput>& output, size_t ancestors, size_t descendants);
     bool EligibleForSpending(const CoinEligibilityFilter& eligibility_filter) const;
     CAmount GetSelectionAmount() const;
 };
 
+struct Groups {
```

Comment:
I believe the separation between positive and mixed group is an implementation detail of particular coin selection algorithms. We can contrast it with `CoinEligibilityFilter` and `OutputType` grouping which are algorithm independent. Therefore I propose to not separate positive outputs at the grouping stage.

Rather filter negative/positive groups  as close to the particular coin selection algo as possible, e.g. in `ChooseSelectionResult`. This will significantly simplify the code by removing extra nesting of structs and multiple `if` statements. I expect computation overhead of iterating over groups to remove negative UTXos to be negligible, but maybe I'm wrong.

Reply:
hmm, that would make the avoid partial spends flow even uglier, plus I don't think that that will be negligible for big wallets (+50k UTXOs wallets).

The later removal of negative UTXOs will leave APS full groups with less than 100 entries, which will force us to re-order groups so they are full again (moving UTXOs from one group to another), then if the wallet does not include partial groups, will need to discard those groups too.
So making this changes would actually introduce more code than what we currently have with few nested structs whose creation is scoped only to `GroupOutputs`. I mean, we would need to have one APS procedure inside `GroupOutputs` and another APS procedure inside `ChooseSelectionResult`.

About the computation overhead:
In the worst case scenario, the wallet executes `ChooseSelectionResult` 5 times in each of the 8 `AttemptSelection` calls.
So, it could end up performing the traverse + re-order + removal 40 times.

Reply:
I'm not sure this is how APS works. When we remove negative UTXOs we assess the whole group and not each UTXO individually, so you don't need to regroup anything after filtering negative groups.

Reply:
> I'm not sure this is how APS works. When we remove negative UTXOs we assess the whole group and not each UTXO individually, so you don't need to regroup anything after filtering negative groups.

No wait, that would leave negative UTXOs in positive-only groups if the sum of the positive UTXOs surpasses the sum of the negative ones.

Currently, a "positive-only group" is essentially a container who stores only positive UTXOs. Not a mix of UTXOs which values sum is positive. That is different conceptually and something that, at least for now, we are not allowing.

Check master, inside `OutputGroup::Insert` method, first line. We reject individual negative UTXOs when the container only stores positive-only UTXOs.

The only way to create positive-groups out of mixed groups is to traverse each group UTXOs set and remove the negative ones. Then, if the group has less than a 100 entries and there are other groups corresponding to the same script (after doing the same removal on those groups), the wallet need to re-order them so only full groups are passed to the selection algorithms (if the avoid partial groups flag is enabled).

Aside from that, performance wise, I still think that performing the negative UTXOs removal later on the process will introduce a non-negligible overhead in big wallets.

Reply:
You're right, thanks for the explanation. This comment could be resolved.